lib/icinga/dependency.cpp

```diff
@@ -9,106 +9,163 @@
+				} else {
+					s << " (implicit)";
+				}
 			}
-
-			oss << "\n-> ";
+			BOOST_THROW_EXCEPTION(ScriptError(s.str()));
 		}
 
-		oss << parent->GetReflectionType()->GetName() << " '" << parent->GetName() << "'";
-
-		if (implicit) {
-			oss << " (implicit)";
+		if (node.Visited) {
+			return;
 		}
-
-		BOOST_THROW_EXCEPTION(ScriptError(oss.str()));
-	}
-
-	AssertNoDependencyCycle(parent, graph, implicit);
-}
-
-static void AssertNoDependencyCycle(const Checkable::Ptr& checkable, DependencyCycleGraph& graph, bool implicit)
-{
-	auto& node (graph.Nodes[checkable]);
-
-	if (!node.Visited) {
 		node.Visited = true;
+
 		node.OnStack = true;
-		graph.Stack.emplace_back(checkable, implicit);
 
-		for (auto& dep : checkable->GetDependencies()) {
-			graph.Stack.emplace_back(dep);
-			AssertNoParentDependencyCycle(dep->GetParent(), graph, false);
-			graph.Stack.pop_back();
+		// Implicit dependency of each service to its host
+		if (auto service (dynamic_pointer_cast<Service>(checkable)); service) {
+			m_Stack.emplace_back(service);
+			AssertNoCycle(service->GetHost());
+			m_Stack.pop_back();
 		}
 
-		{
-			auto service (dynamic_pointer_cast<Service>(checkable));
+		// Explicitly configured dependency objects
+		for (const auto& dep : checkable->GetDependencies()) {
+			m_Stack.emplace_back(dep);
+			AssertNoCycle(dep->GetParent());
+			m_Stack.pop_back();
+		}
 
-			if (service) {
-				AssertNoParentDependencyCycle(service->GetHost(), graph, true);
-			}
+		// Additional dependencies to consider
+		for (const auto& dep : node.ExtraDependencies) {
+			m_Stack.emplace_back(dep);
+			AssertNoCycle(dep->GetParent());
+			m_Stack.pop_back();
 		}
 
-		graph.Stack.pop_back();
 		node.OnStack = false;
 	}
-}
+};
 
-void Dependency::AssertNoCycles()
+/**
+ * Checks that adding these new dependencies to the configuration does not introduce any cycles.
+ *
+ * This is done as an optimization: cycles are checked once for all dependencies in a batch of config objects instead
+ * of individually per dependency in Dependency::OnAllConfigLoaded(). For runtime updates, this function may still be
+ * called for single objects.
+ *
+ * @param items Config items containing Dependency objects added to the running configuration.
+ */
+void Dependency::BeforeOnAllConfigLoadedHandler(const ConfigItems& items)
 {
-	DependencyCycleGraph graph;
-
-	for (auto& host : ConfigType::GetObjectsByType<Host>()) {
-		AssertNoDependencyCycle(host, graph);
-	}
-
-	for (auto& service : ConfigType::GetObjectsByType<Service>()) {
-		AssertNoDependencyCycle(service, graph);
-	}
-
-	m_AssertNoCyclesForIndividualDeps = true;
+	DependencyCycleChecker checker;
+
+	// Resolve parent/child names to Checkable::Ptr and temporarily add the edges to the checker.
+	// The dependencies are later registered to the checkables by Dependency::OnAllConfigLoaded().
+	items.ForEachObject<Dependency>([&checker](Dependency::Ptr dependency) {
```

Comment:
```suggestion
	items.ForEachObject<Dependency>([&checker](Ptr dependency) {
```

Reply:
Possible to do, but doesn't help with readability in my opinion.